### PR TITLE
[BP-62] Bookkeeper client introduce batch read request api.

### PR DESCRIPTION
**Pull request by horizonzy**
Descriptions of the changes in this PR:
This is the third PR for the batch read(https://github.com/apache/bookkeeper/pull/4051) feature.

Bookkeeper client introduces batch read request API.

### CODE REVIEW

**Comment by horizonzy** [comment]
rerun failure checks

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieProtocol.java [hunk 1]
```diff
@@ -562,6 +562,10 @@ class BatchedReadResponse extends Response implements ReferenceCounted {
         final long requestId;
         final ByteBufList data;
 
+        BatchedReadResponse(byte protocolVersion, int errorCode, long ledgerId, long entryId, long requestId) {
+            this(protocolVersion, errorCode, ledgerId, entryId, requestId, ByteBufList.get());
```
Comment (hangc0276) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieProtocol.java:566]: Reminder: No matter what the error code is, don't forget to release the ByteBufList in the client side.

**Review by hangc0276 — APPROVED** [review]

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieProtocol.java (hunk 1 above)
Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieProtocol.java:566]: Sure, the error BatchedReadResponse will also be released.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/proto/MockBookies.java [hunk 2]
```diff
@@ -123,6 +123,47 @@ public ByteBuf readEntry(BookieId bookieId, int flags, long ledgerId, long entry
         return entry;
     }
 
+    public ByteBufList batchReadEntry(BookieId bookieId, int flags, long ledgerId, long startEntryId,
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/proto/MockBookies.java:126]: nit: batchReadEntries ? (plural)

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/proto/MockBookies.java [hunk 3]
```diff
@@ -123,6 +123,47 @@ public ByteBuf readEntry(BookieId bookieId, int flags, long ledgerId, long entry
...
+                if (data == null) {
+                    data = ByteBufList.get(entry);
+                } else {
+                    if (frameSize > maxSize) {
+                        entry.release();
+                        break;
+                    }
+                    data.add(entry);
+                }
+            } catch (Throwable e) {
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/proto/MockBookies.java:157]: catching Throwable is always a code smell, why ? what do you want to catch here ?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java [hunk 4]
```diff
@@ -687,7 +688,7 @@ void connectIfNeededAndDoOp(GenericCallback<PerChannelBookieClient> op) {
     void writeLac(final long ledgerId, final byte[] masterKey, final long lac, ByteBufList toSend, WriteLacCallback cb,
             Object ctx) {
         final long txnId = getTxnId();
-        final CompletionKey completionKey = new V3CompletionKey(txnId,
+        final CompletionKey completionKey = new TxnCompletionKey(txnId,
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java:691]: why did you change this name ?V3CompletionKey > TxnCompletionKey

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java [hunk 5]
```diff
@@ -2340,28 +2467,28 @@ private long getTxnId() {
         return txnIdGenerator.incrementAndGet();
     }
 
-    private final Recycler<V2CompletionKey> v2KeyRecycler = new Recycler<V2CompletionKey>() {
+    private final Recycler<EntryCompletionKey> v2KeyRecycler = new Recycler<EntryCompletionKey>() {
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java:2470]: why are you changing this name ? V2CompletionKey > EntryCompletionKey
it will make porting patches to other branches harder.
if it is not strictly needed I won't change the class names

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java (hunk 4 above)
Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java:691]: The `V3CompletionKey` may mislead. In this PR, it's used for the v2 batch read. The V3CompletionKey uses txnId to trace, so I change the name to TxnCompletionKe.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java (hunk 5 above)
Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java:2470]: We are renaming V3CompletionKey to TxnCompletionKey. Therefore, I want to downplay the concepts of V2 and V3 and name it based on its purpose.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/proto/MockBookies.java (hunk 3 above)
Comment (horizonzy) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/proto/MockBookies.java:157]: Make sense, I copy the code from the BatchedReadEntryProcessor, here we don't need catch the exception.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/proto/MockBookies.java (hunk 2 above)
Comment (horizonzy) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/proto/MockBookies.java:126]: make sense.

**Review by eolivelli — APPROVED** [review]
LGTM
